tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11236,1074 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+            TestConfig.OperatorString.c_str());
+    } else // Unary op logic
+    {
+      if (TestConfig.IntrinsicString == "testClamp")
+        VecExpected.push_back(
+            std::clamp(VecInput1[Index], ClampArgC, ClampArgT));
+      else if (TestConfig.IntrinsicString == "testInitialize")
+        VecExpected.push_back(VecInput1[Index]);
+      else
+        LogErrorFmt(L"Unrecognized intrinsic string: %s",
+                    TestConfig.IntrinsicString.c_str());
+    }
+  }
+
+  // 1,2, and 3 are required to compile the shader.
+  // 1. -DOPERATOR : "*" "+" "," "-" "/" etc.
+  // 2. -DOPERAND2: : l.VecInput2 or scalarInput, depending on the test.
+  // 3. -DCLAMP_ARGC, -DCLAMP_ARG2 : clamp arg 1 and 2. Required to compile.
+  // 4. Optional: -DFUNC : min, max, etc.
+  std::string ShaderText = R"(
+  struct LongVectorOp{
+    TYPE ScalarInput;
+    vector<TYPE, NUM> VecInput1;
+    vector<TYPE, NUM> VecInput2;
+    vector<TYPE, NUM> VecOutput;
+  };
+
+  vector<TYPE, NUM> testInitialize(vector<TYPE, NUM> Vec)
+  {
+    vector<TYPE, NUM> VecCopy = Vec;
+    return VecCopy;
+  }
+
+  vector<TYPE, NUM> testClamp(vector<TYPE, NUM> Vec)
+  {
+    return clamp(Vec, TYPE(CLAMP_ARGC), TYPE(CLAMP_ARGT));
+  }
+
+  RWStructuredBuffer<LongVectorOp> g_buf : register(u0);
+  RWByteAddressBuffer g_bufByteAddr : register(u1);
+  [numthreads(8,8,1)]
+  void main(uint GI : SV_GroupIndex) {
+    LongVectorOp LVOp = g_buf[GI];
+    LVOp.VecOutput = FUNC(LVOp.VecInput1 OPERATOR OPERAND2);
+    g_buf[GI] = LVOp;
+  };
+  )";
+
+  // Set up the compiler options string.
+  std::stringstream CompilerOptions("");
+  std::string HLSLType = TestConfig.GetHLSLTypeString();
+  CompilerOptions << "-DTYPE=";
+  CompilerOptions << HLSLType;
+  CompilerOptions << " -DNUM=";
+  CompilerOptions << N;
+  const bool Is16BitType =
+      (HLSLType == "int16_t" || HLSLType == "uint16_t" || HLSLType == "half");
+  CompilerOptions << (Is16BitType ? " -enable-16bit-types" : "");
+  CompilerOptions << " -DOPERATOR=";
+  CompilerOptions << TestConfig.OperatorString;
+  CompilerOptions << " -DOPERAND2=";
+  if (TestConfig.IsBinaryOp) {
+    CompilerOptions << (TestConfig.IsScalarOp ? "LVOp.ScalarInput"
+                                              : "LVOp.VecInput2");
+  }
+  CompilerOptions << " -DFUNC=";
+  CompilerOptions << TestConfig.IntrinsicString;
+  CompilerOptions << " -DCLAMP_ARGC=";
+  CompilerOptions << ClampArgC;
+  CompilerOptions << " -DCLAMP_ARGT=";
+  CompilerOptions << ClampArgT;
+
+  // We have to construct the string outside of the lambda. Otherwise it's
+  // cleaned up when the lambda finishes executing but before the shader runs.
+  std::string CompilerOptionsString = CompilerOptions.str();
+
+  // ShaderOpArith.xml defines some boiler plate for the shader code.
+  CComPtr<IStream> TestXML;
+  ReadHlslDataIntoNewStream(L"ShaderOpArith.xml", &TestXML);
+
+  // RunShaderOpTest is a helper function that does all device/resource creation
```

Comment:
Ah, and that's pre-existing, so there's not much point digging into why it returns a shared_ptr rather than a unique_ptr.